message-broker/src/lib.rs

```diff
@@ -27,57 +27,112 @@ use frame_support::{
 	StorageValue,
 };
-use frame_system::ensure_none;
+use frame_system::{ensure_none, ensure_root};
 use sp_inherents::{InherentData, InherentIdentifier, MakeFatalError, ProvideInherent};
 use sp_std::{cmp, prelude::*};
 
 use cumulus_primitives::{
-	inherents::{DownwardMessagesType, DOWNWARD_MESSAGES_IDENTIFIER},
-	well_known_keys, DownwardMessageHandler, InboundDownwardMessage, UpwardMessage,
+	inherents::{MessageIngestionType, MESSAGE_INGESTION_IDENTIFIER},
+	well_known_keys, DownwardMessageHandler, HrmpMessageHandler, OutboundHrmpMessage,
+	UpwardMessage, ParaId,
 };
 
+// TODO: these should be not a constant, but sourced from the relay-chain configuration.
+const UMP_MSG_NUM_PER_CANDIDATE: usize = 5;
+const HRMP_MSG_NUM_PER_CANDIDATE: usize = 5;
+
 /// Configuration trait of the message broker pallet.
 pub trait Config: frame_system::Config {
 	/// The downward message handlers that will be informed when a message is received.
 	type DownwardMessageHandlers: DownwardMessageHandler;
+	/// The HRMP message handlers that will be informed when a message is received.
+	type HrmpMessageHandlers: HrmpMessageHandler;
 }
 
 decl_storage! {
 	trait Store for Module<T: Config> as MessageBroker {
 		PendingUpwardMessages: Vec<UpwardMessage>;
+
+		/// Essentially `OutboundHrmpMessage`s grouped by the recipients.
+		OutboundHrmpMessages: map hasher(twox_64_concat) ParaId => Vec<Vec<u8>>;
+		/// HRMP channels with the given recipients are awaiting to be processed. If a `ParaId` is
+		/// present in this vector then `OutboundHrmpMessages` for it should be not empty.
+		NonEmptyHrmpChannels: Vec<ParaId>;
 	}
 }
 
 decl_module! {
 	pub struct Module<T: Config> for enum Call where origin: T::Origin {
 		/// An entrypoint for an inherent to deposit downward messages into the runtime. It accepts
-		/// and processes the list of downward messages.
+		/// and processes the list of downward messages and inbound HRMP messages.
 		#[weight = (10, DispatchClass::Mandatory)]
-		fn receive_downward_messages(origin, messages: Vec<InboundDownwardMessage>) {
+		fn ingest_inbound_messages(origin, messages: MessageIngestionType) {
 			ensure_none(origin)?;
 
-			let messages_len = messages.len() as u32;
-			for message in messages {
-				T::DownwardMessageHandlers::handle_downward_message(message);
+			let MessageIngestionType {
+				dmp,
+				hrmp,
+			} = messages;
+
+			let dmp_count = dmp.len() as u32;
+			for dmp_message in dmp {
+				T::DownwardMessageHandlers::handle_downward_message(dmp_message);
 			}
 
 			// Store the processed_downward_messages here so that it's will be accessible from
 			// PVF's `validate_block` wrapper and collation pipeline.
 			storage::unhashed::put(
 				well_known_keys::PROCESSED_DOWNWARD_MESSAGES,
-				&messages_len,
+				&dmp_count,
 			);
+
+			let mut hrmp_watermark = None;
+			for (sender, channel_contents) in hrmp {
+				for hrmp_message in channel_contents {
+					hrmp_watermark = Some(hrmp_message.sent_at);
```

Comment:
Should we not check that we only update the watermark if `sent_at` is greater?

Reply:
well, yes and no.

Here, we rely on a property of a channel, that each message has `sent_at` always strictly greater than the last one. So it's this should be fine. However, we are not checking right now if the collator submits the right messages. In fact, it can submit any messages and get away with it. That's the concern for #235 though.

I can add an additional check but then with #235 it will get redundant so I'd rather not.

Reply:
But we iterate here over all the channels? 

Can there not exist a channel that has `sent_at` being lower than the others?

Reply:
oh got it now, yes you are right